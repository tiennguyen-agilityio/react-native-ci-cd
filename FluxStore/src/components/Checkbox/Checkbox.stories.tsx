import {useState} from 'react';
import {Meta, StoryObj} from '@storybook/react';

import Checkbox, {CheckboxProps} from '.';

const meta: Meta<typeof Checkbox> = {
  title: 'Components/Checkbox',
  component: Checkbox,
};

export default meta;

type Story = StoryObj<typeof Checkbox>;

type RenderProps = Omit<CheckboxProps, 'onValueChange'>;

const Render = ({selected, ...props}: RenderProps) => {
  const [isChecked, setIsChecked] = useState(selected);
  const handleChange = () => setIsChecked(prev => !prev);
  return <Checkbox selected={isChecked} onValueChange={handleChange} {...props} />;
};

export const Default: Story = {
  render: Render,
};

export const Checked: Story = {
  render: () => <Render selected label="Label checkbox" />,
};

export const Disabled: Story = {
  render: () => <Render disabled label="Label checkbox" />,
};

export const DisabledChecked: Story = {
  render: () => <Render disabled selected label="Label checkbox" />,
};
