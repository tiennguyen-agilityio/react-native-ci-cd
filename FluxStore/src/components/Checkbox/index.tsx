import {memo, useMemo} from 'react';
import Svg, {Path} from 'react-native-svg';
import {View, Text, ViewStyle, TouchableOpacity, StyleSheet, TextStyle} from 'react-native';

// Stores
import {useThemeStore} from '@/stores';
import {colors, fontFamilies, fontSizes} from '@/themes';

const styles = StyleSheet.create({
  wrapper: {
    flexDirection: 'row',
    alignItems: 'center',
    marginBottom: 20,
    gap: 14,
  },
});

export interface CheckboxProps {
  label?: string;
  size?: number;
  selected?: boolean;
  disabled?: boolean;
  onValueChange: () => void;
}

const Checkbox = ({
  label = '',
  size = 20,
  disabled = false,
  selected = false,
  onValueChange,
}: CheckboxProps) => {
  const {theme} = useThemeStore();

  const checkboxStyles: ViewStyle = useMemo(() => {
    let borderColor = selected ? theme.border.quinary : theme.border.quaternary;
    let backgroundColor = selected ? theme.background.checkbox : theme.transparent;

    if (disabled) {
      borderColor = colors.gray.A500;
      backgroundColor = selected ? colors.gray.A500 : colors.transparent;
    }

    return {
      borderWidth: 1,
      alignItems: 'center',
      justifyContent: 'center',
      width: size,
      height: size,
      borderColor,
      backgroundColor,
    };
  }, [
    size,
    selected,
    disabled,
    theme.transparent,
    theme.border.quinary,
    theme.border.quaternary,
    theme.background.checkbox,
  ]);

  const labelStyles: TextStyle = useMemo(
    () => ({
      fontFamily: fontFamilies.productSans.medium,
      fontSize: fontSizes.xs,
      color: disabled ? colors.gray.A500 : theme.text.default,
    }),
    [theme.text.default, disabled],
  );

  return (
    <TouchableOpacity
      activeOpacity={0.8}
      disabled={disabled}
      onPress={onValueChange}
      style={styles.wrapper}>
      <View style={checkboxStyles}>
        {selected && (
          <Svg width={9} height={7} viewBox="0 0 9 7" fill="none">
            <Path
              d="M3.535 7a.995.995 0 01-.711-.3L.292 4.116a1.037 1.037 0 010-1.442.987.987 0 011.413 0l1.802 1.839L7.262.33A.98.98 0 018.675.273c.408.377.437 1.026.057 1.442L4.275 6.671A.974.974 0 013.554 7h-.019z"
              fill={colors.white[500]}
            />
          </Svg>
        )}
      </View>
      {label && <Text style={labelStyles}>{label}</Text>}
    </TouchableOpacity>
  );
};

export default memo(Checkbox);
