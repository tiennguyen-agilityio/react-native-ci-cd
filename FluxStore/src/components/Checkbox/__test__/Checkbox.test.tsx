import {render, fireEvent, screen} from '@testing-library/react-native';
import Checkbox from '..';

const onValueChange = jest.fn();

describe('Checkbox', () => {
  const props = {
    selected: false,
    label: 'Label checkbox',
    onValueChange,
  };

  it('renders with label', () => {
    const {toJSON} = render(<Checkbox {...props} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders with no label', () => {
    const {toJSON} = render(<Checkbox onValueChange={onValueChange} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders with selected false', () => {
    const {toJSON} = render(<Checkbox {...props} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders with selected true', () => {
    const {toJSON} = render(<Checkbox {...props} selected={true} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders with selected true, disabled true', () => {
    const {toJSON} = render(<Checkbox {...props} selected disabled />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders with disabled true', () => {
    const {toJSON} = render(<Checkbox {...props} disabled />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('calls onValueChange when pressed', () => {
    render(<Checkbox label="Click me" onValueChange={onValueChange} />);

    fireEvent.press(screen.getByText('Click me'));

    expect(onValueChange).toHaveBeenCalled();
  });
});
