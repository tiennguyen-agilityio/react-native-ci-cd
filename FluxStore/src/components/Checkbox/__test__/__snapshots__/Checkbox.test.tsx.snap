// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox renders with disabled true 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": true,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "flexDirection": "row",
      "gap": 14,
      "marginBottom": 20,
      "opacity": 1,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "transparent",
        "borderColor": "#bebfc4",
        "borderWidth": 1,
        "height": 20,
        "justifyContent": "center",
        "width": 20,
      }
    }
  />
  <Text
    style={
      {
        "color": "#bebfc4",
        "fontFamily": "ProductSansMedium",
        "fontSize": 14,
      }
    }
  >
    Label checkbox
  </Text>
</View>
`;

exports[`Checkbox renders with label 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "flexDirection": "row",
      "gap": 14,
      "marginBottom": 20,
      "opacity": 1,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "transparent",
        "borderColor": "#bebfc4",
        "borderWidth": 1,
        "height": 20,
        "justifyContent": "center",
        "width": 20,
      }
    }
  />
  <Text
    style={
      {
        "color": "#000",
        "fontFamily": "ProductSansMedium",
        "fontSize": 14,
      }
    }
  >
    Label checkbox
  </Text>
</View>
`;

exports[`Checkbox renders with no label 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "flexDirection": "row",
      "gap": 14,
      "marginBottom": 20,
      "opacity": 1,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "transparent",
        "borderColor": "#bebfc4",
        "borderWidth": 1,
        "height": 20,
        "justifyContent": "center",
        "width": 20,
      }
    }
  />
</View>
`;

exports[`Checkbox renders with selected false 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "flexDirection": "row",
      "gap": 14,
      "marginBottom": 20,
      "opacity": 1,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "transparent",
        "borderColor": "#bebfc4",
        "borderWidth": 1,
        "height": 20,
        "justifyContent": "center",
        "width": 20,
      }
    }
  />
  <Text
    style={
      {
        "color": "#000",
        "fontFamily": "ProductSansMedium",
        "fontSize": 14,
      }
    }
  >
    Label checkbox
  </Text>
</View>
`;

exports[`Checkbox renders with selected true 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "flexDirection": "row",
      "gap": 14,
      "marginBottom": 20,
      "opacity": 1,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#5ece7b",
        "borderColor": "#5ece7b",
        "borderWidth": 1,
        "height": 20,
        "justifyContent": "center",
        "width": 20,
      }
    }
  >
    <RNSVGSvgView
      align="xMidYMid"
      bbHeight={7}
      bbWidth={9}
      fill="none"
      focusable={false}
      height={7}
      meetOrSlice={0}
      minX={0}
      minY={0}
      style={
        [
          {
            "backgroundColor": "transparent",
            "borderWidth": 0,
          },
          {
            "flex": 0,
            "height": 7,
            "width": 9,
          },
        ]
      }
      vbHeight={7}
      vbWidth={9}
      width={9}
    >
      <RNSVGGroup
        fill={null}
        propList={
          [
            "fill",
          ]
        }
      >
        <RNSVGPath
          d="M3.535 7a.995.995 0 01-.711-.3L.292 4.116a1.037 1.037 0 010-1.442.987.987 0 011.413 0l1.802 1.839L7.262.33A.98.98 0 018.675.273c.408.377.437 1.026.057 1.442L4.275 6.671A.974.974 0 013.554 7h-.019z"
          fill={
            {
              "payload": 4294967295,
              "type": 0,
            }
          }
          propList={
            [
              "fill",
            ]
          }
        />
      </RNSVGGroup>
    </RNSVGSvgView>
  </View>
  <Text
    style={
      {
        "color": "#000",
        "fontFamily": "ProductSansMedium",
        "fontSize": 14,
      }
    }
  >
    Label checkbox
  </Text>
</View>
`;

exports[`Checkbox renders with selected true, disabled true 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": true,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "flexDirection": "row",
      "gap": 14,
      "marginBottom": 20,
      "opacity": 1,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#bebfc4",
        "borderColor": "#bebfc4",
        "borderWidth": 1,
        "height": 20,
        "justifyContent": "center",
        "width": 20,
      }
    }
  >
    <RNSVGSvgView
      align="xMidYMid"
      bbHeight={7}
      bbWidth={9}
      fill="none"
      focusable={false}
      height={7}
      meetOrSlice={0}
      minX={0}
      minY={0}
      style={
        [
          {
            "backgroundColor": "transparent",
            "borderWidth": 0,
          },
          {
            "flex": 0,
            "height": 7,
            "width": 9,
          },
        ]
      }
      vbHeight={7}
      vbWidth={9}
      width={9}
    >
      <RNSVGGroup
        fill={null}
        propList={
          [
            "fill",
          ]
        }
      >
        <RNSVGPath
          d="M3.535 7a.995.995 0 01-.711-.3L.292 4.116a1.037 1.037 0 010-1.442.987.987 0 011.413 0l1.802 1.839L7.262.33A.98.98 0 018.675.273c.408.377.437 1.026.057 1.442L4.275 6.671A.974.974 0 013.554 7h-.019z"
          fill={
            {
              "payload": 4294967295,
              "type": 0,
            }
          }
          propList={
            [
              "fill",
            ]
          }
        />
      </RNSVGGroup>
    </RNSVGSvgView>
  </View>
  <Text
    style={
      {
        "color": "#bebfc4",
        "fontFamily": "ProductSansMedium",
        "fontSize": 14,
      }
    }
  >
    Label checkbox
  </Text>
</View>
`;
